feat: Added Script Window

GustavoLR548 opened the pull request:
# Script Window for generated GDScript code

Instead of just printing the results to the screen, I created a Window to popup with the generated script. 

You can also copy the entire generated code with a press of a button

If there is any problem with the code or anything you would like to change, please let me know!

Edit: The "Copiar" is misleading, I inserted the text as "Copy", but when running the code, it translated to "Copiar" for some reason that I cannot explain.

![image](https://github.com/user-attachments/assets/912b3c42-09aa-4c1e-89f9-6f7809b4cb3c)

dbnicholson commented on addons/block_code/plugin.cfg:
```diff
@@ -3,5 +3,5 @@
 description="Create games using a high-level, block-based visual programming language."
 author="Endless"
-version="$Format:%(describe:tags)$"
+version="v0.6.1-27-g73861397"
```
Can you leave this out of the patch? It's there for when we make releases.

dbnicholson commented on addons/block_code/ui/main_panel.gd:
```diff
@@ -59,7 +61,12 @@ func _on_undo_redo_version_changed():
+	
+	#BlockCodePlugin.script_window.script_content = script
+	add_script_window.emit(script)
+	#_print_generated_script()
```
Can you remove the commented lines?

dbnicholson commented on addons/block_code/ui/main_panel.gd:
```diff
@@ -1,6 +1,8 @@
 @tool
 extends Control
 
+signal add_script_window(script: String)
```
Can we call this signal `script_window_requested`?

dbnicholson commented on addons/block_code/ui/script_window/script_window.gd:
```diff
@@ -0,0 +1,19 @@
+
+
+func _on_close_requested():
+	visible = false
```
I don't think this is ever connected, so I think it can be removed.

dbnicholson reviewed (changes requested):
I like this and had been thinking about doing the same thing for a while. I left a few cleanup comments, but overall this looks good to me. Thanks!

dbnicholson commented on addons/block_code/ui/script_window/script_window.tscn:
```diff
@@ -0,0 +1,63 @@
+layout_mode = 2
+size_flags_horizontal = 0
+size_flags_vertical = 0
+text = "Here is what your Block Code looks like in GDScript!"
```
I think this would be better as the window title instead of a label in the window. I think I'd probably make it something a little more generic like `Block Code Generated GDScript`. But we can always fine tune it after merging.

GustavoLR548 commented:
Happy to see you liked the feature :).

Sure thing! I will make the requested changes.

GustavoLR548 commented:
If there is any other problem with the pull request, let me know ;-)

manuq commented on addons/block_code/ui/script_window/script_window.gd:
```diff
@@ -0,0 +1,14 @@
+@tool
+extends Window
+class_name ScriptWindow
```
@GustavoLR548 please remove the class name. You are already preloading this script fine in `addons/block_code/block_code_plugin.gd` using the resource path. The problem we found with `class_name` was that they can collide with the user project. Suppose a user has a `ScriptWindow` class in their project and then they add this plugin, the names collide.
```suggestion
```

manuq reviewed (changes requested):
Thank you for the contribution @GustavoLR548 ! One more detail.

GustavoLR548 commented:
Okay, I removed the class name.

Additionally, I changed the name of the node and the function called by the signal responsible for showing the code to reflect the changes made.

Let me know if you need any further adjustments.

manuq commented on addons/block_code/block_code_plugin.gd:
```diff
@@ -59,6 +61,17 @@ func _enter_tree():
+func script_window_requested(script: String):
+	
+	var script_window = ScriptWindow.instantiate()
+	script_window.script_content = script
```
@GustavoLR548 I noticed in your screenshot that there is no indentation. I think you can get indentation by replacing the tabs with 4 spaces:
```suggestion
	script_window.script_content = script.replace('\t', '    ')
```

And then add a monospace font to the label:
- Add Label Settings
- Add Font -> SystemFont
- Font Names -> Add Element -> Monospace
![Captura desde 2024-08-01 13-55-34](https://github.com/user-attachments/assets/0165d465-47da-45fd-a34f-1606ae9ec898)

The result will be something like this:
![Captura desde 2024-08-01 14-00-30](https://github.com/user-attachments/assets/db74838a-37bd-489b-8938-24f0d9053f51)

GustavoLR548 commented:
Here, not only I changed the font, but also discovered that you can add a syntax highlighter to the text, and so I decided to add :D

![image](https://github.com/user-attachments/assets/4adc3591-d54b-46ee-8a7b-dc7f00d499a6)

GustavoLR548 commented:
Ok, added the Monospace font. Also there is now a syntax highlighter :D

![image](https://github.com/user-attachments/assets/3eaae7dd-13d6-453c-91d1-061ffce540b4)

manuq reviewed (approved):
@GustavoLR548 with syntax highlight this is just perfect! Please squash your hotfixes into a single commit so we can merge.

GustavoLR548 commented:
Perfect! Just to inform I send another commit to add color to more keywords that were not present before.

![image](https://github.com/user-attachments/assets/fd8e6c28-4d64-47ea-ba82-7b9e61c46bd6)

Ok, I will squash the hotfixes

manuq commented:
> Ok, I think I was able to correctly squash the commits into a single one.
> 
> If I made any mistakes or if there is any other adjustments to be made, let me know

Obrigado! This is now in a single commit. As a last request, can I ask you to change the commit description to describe the change? Something like the following would be enough:

> Show generated script in a window
>
> Replace printing the generated script to the output. Display it in a separate window instead.

GustavoLR548 commented:
De nada :)

Sure! I will change the commit description.

manuq commented:
@GustavoLR548 your contribution is now merged and will be in next release, thank you! We have some checks and they were failing so I decided to fix them myself in #183 .

dbnicholson commented:
In the future it might be best to have the contributor [allow edits by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) and update their branch directly. Then you won't need to open a separate PR on a separate branch.

manuq commented:
> In the future it might be best to have the contributor [allow edits by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) and update their branch directly. Then you won't need to open a separate PR on a separate branch.

Oh I didn't know about that feature. Thanks @dbnicholson, I will do so next time.
